### PR TITLE
Removed auto-decoding from the floating point version of std.conv.parse

### PR DESCRIPTION
**Pull request by JackStouffer**
New version of https://github.com/dlang/phobos/pull/4674

13.5x faster! [Code here](https://gist.github.com/JackStouffer/0cf93d3cc861d699b334e18ea67fb66c#file-test-d).

```
# new
$ ldc2 -O5 -release test.d && ./test
Total time: 387 ms, 506 μs, and 8 hnsecs

# old
$ ldc2 -O5 -release test.d && ./test
Total time: 5 secs, 238 ms, 59 μs, and 4 hnsecs
```

### CODE REVIEW

**Comment by JackStouffer** [comment]
ping @andralex

File: std/conv.d [hunk 1]
```diff
@@ -2541,34 +2541,48 @@ Target parse(Target, Source)(ref Source p)
...
-    switch (p.front)
+    bool sign = false;
+    switch (source.front)
     {
     case '-':
-        sign++;
-        p.popFront();
-        enforce(!p.empty, bailOut());
-        if (toLower(p.front) == 'i')
+        sign = true;
```
Comment (andralex) [inline std/conv.d]: So I guess an `enforce(!sign)` is in order here. In fact `sign` should be a ternary: not seen, `+` seen, `-` seen.

File: std/conv.d [hunk 2]
```diff
@@ -2541,34 +2541,48 @@ Target parse(Target, Source)(ref Source p)
...
+        source.popFront();
```
Comment (andralex) [inline std/conv.d]: (For simpler refactoring: continue using `p` and give the initial parameter a different name.)

File: std/conv.d [hunk 3]
```diff
@@ -2577,16 +2591,19 @@ Target parse(Target, Source)(ref Source p)
...
     if (startsWithZero)
     {
-        p.popFront();
-        if (p.empty)
+        source.popFront();
+        if (source.empty)
         {
+            static if (isNarrowString!Source)
+                p = source.assumeUTF;
+
```
Comment (andralex) [inline std/conv.d]: no reason for vertical space here

File: std/conv.d [hunk 4]
```diff
@@ -2577,16 +2591,19 @@ Target parse(Target, Source)(ref Source p)
...
             return (sign) ? -0.0 : 0.0;
```
Comment (andralex) [inline std/conv.d]: drop the parens

Comment (andralex) [inline std/conv.d]: cc @wilzbach :)

File: std/conv.d [hunk 5]
```diff
@@ -2577,16 +2591,19 @@ Target parse(Target, Source)(ref Source p)
...
         }
 
-        isHex = p.front == 'x' || p.front == 'X';
+        isHex = source.front == 'x' || source.front == 'X';
```
Comment (andralex) [inline std/conv.d]: `source.front.among('x', 'X')` if it's convenient

File: std/conv.d [hunk 6]
```diff
@@ -2668,20 +2685,20 @@ Target parse(Target, Source)(ref Source p)
         }
 
         enforce(anydigits, bailOut());
-        enforce(!p.empty && (p.front == 'p' || p.front == 'P'),
+        enforce(!source.empty && (source.front == 'p' || source.front == 'P'),
```
Comment (andralex) [inline std/conv.d]: among could be of use again

**Comment by JackStouffer** [comment]
When back to `p` so the diff is now much smaller

**Comment by JackStouffer** [comment]
Auto-merge toggled on

**Comment by jondegenhardt** [comment]
Very nice! This should improve performance of conversions by std.conv.to from string to float or double, right? If that's the case, it might be worth including the std.conv.to benefit in the update notes. Might not be obvious.

**Pull request merged**